client/network/sync/src/lib.rs

```diff
@@ -601,14 +602,27 @@ where
 				if let SyncMode::Warp = &self.mode {
 					if self.peers.len() >= MIN_PEERS_TO_START_WARP_SYNC && self.warp_sync.is_none()
 					{
-						log::debug!(target: "sync", "Starting warp state sync.");
-						if let Some(provider) = &self.warp_sync_provider {
-							self.warp_sync =
-								Some(WarpSync::new(self.client.clone(), provider.clone()));
+						match self.warp_sync_params.as_mut() {
+							Some(WarpSyncParams::WithProvider(warp_with_provider)) => {
+								log::debug!(target: "sync", "Starting warp state sync.");
+								self.warp_sync = Some(WarpSync::new(
+									self.client.clone(),
+									warp_with_provider.clone(),
+								));
+							},
+							Some(WarpSyncParams::WaitForTarget(target_block)) => {
+								log::debug!(target: "sync", "Waiting for target block.");
+								futures::executor::block_on(async {
+									self.warp_sync = Some(WarpSync::new_with_target_block(
+										self.client.clone(),
+										target_block.await.unwrap(),
+									));
+								});
```

Comment:
You should not block here!

In the internals of `WarpSync` you would need to poll the sender until the target block was send.

Reply:
Thanks Basti, I thought `WarpSync` should only receive the header once it's known.

Given both `ChainSync` and `WarpSync` are synchronous, is there a way to poll synchronously without blocking?

Reply:
Looking at the implementation of [Receiver](https://docs.rs/futures/0.1.19/futures/sync/oneshot/struct.Receiver.html), it doesn't look like there is a way to synchronously poll. Based on that we need to block somewhere? 

I can move the block into the `WarpSync` but I think that's just going to be the same issue no?